kvbc/benchmark/kvbcbench/main.cpp

```diff
@@ -0,0 +1,297 @@
+  auto data = registrar.perf.get("bench");
+  cout << registrar.perf.toString(data) << endl;
+}
+
+std::shared_ptr<rocksdb::Statistics> completeRocksdbConfiguration(
+    ::rocksdb::Options& db_options, std::vector<::rocksdb::ColumnFamilyDescriptor>& cf_descs) {
+  static constexpr size_t CACHE_SIZE = 1024 * 1024 * 1024 * 4ul;  // 4 GB
+  auto table_options = ::rocksdb::BlockBasedTableOptions{};
+  table_options.block_cache = ::rocksdb::NewLRUCache(CACHE_SIZE);
+  table_options.filter_policy.reset(::rocksdb::NewBloomFilterPolicy(10, false));
+  db_options.table_factory.reset(NewBlockBasedTableFactory(table_options));
+
+  // Use the same block cache and table options for all column familes for now.
+  for (auto& d : cf_descs) {
+    auto* cf_table_options =
+        reinterpret_cast<::rocksdb::BlockBasedTableOptions*>(d.options.table_factory->GetOptions());
+    cf_table_options->block_cache = table_options.block_cache;
+    cf_table_options->filter_policy.reset(::rocksdb::NewBloomFilterPolicy(10, false));
+
+    // We never seek on versions
+    if (d.name == categorization::detail::BLOCK_MERKLE_LATEST_KEY_VERSION_CF) {
+      d.options.OptimizeForPointLookup(CACHE_SIZE);
+    }
+  }
+  return db_options.statistics;
+}
+
+std::pair<std::vector<categorization::BlockMerkleInput>, ReadKeys> createBlockInput(const po::variables_map& config) {
+  auto total_blocks = config["total-blocks"].as<size_t>();
+  auto key_size = config["block-merkle-key-size"].as<size_t>();
+  auto value_size = config["block-merkle-value-size"].as<size_t>();
+  auto batch_size = config["batch-size"].as<size_t>();
+  auto num_adds = config["num-block-merkle-keys-add"].as<size_t>() * batch_size;
+  auto num_deletes = config["num-block-merkle-keys-delete"].as<size_t>() * batch_size;
+  auto max_memory_for_kv = config["max-memory-for-kv-gen"].as<size_t>();
+  auto max_read_keys = config["max-total-block-merkle-read-keys"].as<size_t>();
+
+  const auto block_size = (key_size + value_size) * num_adds + key_size * num_deletes;
+  const size_t num_blocks = std::min(max_memory_for_kv / block_size, total_blocks);
+  auto blocks = std::vector<BlockMerkleInput>{};
+  blocks.reserve(num_blocks);
+
+  // Run generation in parallel
+  auto num_cpus = std::thread::hardware_concurrency();
+  cout << "Generating blocks in parallel across " << num_cpus << " cpus" << endl;
+  auto blocks_per_thread = num_blocks / num_cpus;
+
+  // Each thread fills in their subset of keys in a pre-allocated vector. There is no race here, and
+  // no need for locks as the buffer segments in each thread are disjoint.
+  auto total_read_keys = std::min(num_adds * num_blocks, max_read_keys);
+  auto read_keys_per_thread = total_read_keys / num_cpus;
+  auto read_keys = ReadKeys(total_read_keys);
+
+  auto gen = [&](size_t thread_id) -> std::vector<categorization::BlockMerkleInput> {
+    static thread_local std::mt19937 generator;
+    auto distribution = std::uniform_int_distribution<uint8_t>(1, 255);
```

Comment:
Based on https://en.cppreference.com/w/cpp/numeric/random/uniform_int_distribution, it seems if the template parameter is `uint8_t`, result is undefined. Maybe we should use `uint16_t` and cast to `uint8_t`.

```

IntType | - | The result type generated by the generator. The effect is undefined if this is not one
of short, int, long, long long, unsigned short, unsigned int, unsigned long, or unsigned long long
```

Reply:
One more thing - why is 0 excluded?

Reply:
Great catch on both! I don't recall why I didn't include 0. I know in some cases it was to prevent divide by zero, but I don't think that's relevant here.